modules/currency.js

```diff
@@ -0,0 +1,206 @@
+var conversionCache = {};
+var currencyRatesLoaded = false;
+var adServerCurrency = 'USD';
+
+var oldBidResponse;
+
+export var currencySupportEnabled = false;
+export var currencyRates = {};
+
+export function setConfig(config) {
+  var url = DEFAULT_CURRENCY_RATE_URL;
+  if (typeof config.adServerCurrency === 'string') {
+    utils.logInfo('enabling currency support', arguments);
+
+    adServerCurrency = config.adServerCurrency;
+    if (config.conversionRateFile) {
+      utils.logInfo('currency using override conversionRateFile:', config.conversionRateFile);
+      url = config.conversionRateFile;
+    }
+    initCurrency(url);
+  } else {
+    // currency support is disabled, setting defaults
+    utils.logInfo('disabling currency support');
+    resetCurrency();
+  }
+}
+
+function initCurrency(url) {
+  utils.logInfo('Invoking initCurrency', arguments);
+
+  conversionCache = {};
+  currencySupportEnabled = true;
+
+  if(!oldBidResponse) {
+    utils.logInfo('Installing addBidResponse decorator for currency module', arguments);
+
+    oldBidResponse = bidmanager.addBidResponse;
+    bidmanager.addBidResponse = addBidResponseDecorator(bidmanager.addBidResponse);
+  }
+
+  ajax(url, function(response) {
+    try {
+      currencyRates = JSON.parse(response);
+      utils.logInfo('currencyRates set to ' + JSON.stringify(currencyRates));
+    } catch (e) {
+      utils.logError('failed to parse currencyRates response: ' + response);
+    }
+    currencyRatesLoaded = true;
+    processBidResponseQueue();
+  });
+}
+
+function resetCurrency() {
+  if(oldBidResponse) {
+    utils.logInfo('Uninstalling addBidResponse decorator for currency module', arguments);
+
+    bidmanager.addBidResponse = oldBidResponse;
+    oldBidResponse = undefined;
+  }
+
+  adServerCurrency = 'USD';
+  conversionCache = {};
+  currencySupportEnabled = false;
+  currencyRatesLoaded = false;
+  currencyRates = {};
+}
+
+export function addBidResponseDecorator(fn) {
+  return function(adUnitCode, bid) {
+    utils.logInfo('Invoking addBidResponseDecorator function', arguments);
+
+    // default to USD if currency not set
+    if(bid && !bid.currency) {
+      utils.logWarn('Currency not specified on bid.  Defaulted to "USD"');
+      bid.currency = 'USD';
+    }
+
+    if(bid) {
+      // execute immediately if the bid is already in the desired currency
+      if(bid.currency === adServerCurrency) {
+        return fn.apply(this, arguments);
+      }
+
+      bidResponseQueue.push(wrapFunction(fn, this, arguments));
+      if (!currencySupportEnabled || currencyRatesLoaded) {
+        processBidResponseQueue();
+      }
+    }
+  }
+}
+
+function processBidResponseQueue() {
+  utils.logInfo('Invoking processBidResponseQueue', arguments);
+
+  while (bidResponseQueue.length > 0) {
+    (bidResponseQueue.shift())();
+  }
+}
+
+function wrapFunction(fn, context, params) {
+  return function() {
+    utils.logInfo('Invoking wrapped function', arguments);
```

Comment:
not sure how helpful these info statements are as they are deep within internal implementation.